File: src/array/lc76.go
package array

func minWindow(s string, t string) string {
	wndChr2cnt := make(map[byte]int)
	tChr2cnt := make(map[byte]int)
	for _, chr := range t {
		tChr2cnt[byte(chr)]++
	}

	biggerThan := func() bool {
		for chr, cnt := range tChr2cnt {
			if wndCnt, ok := wndChr2cnt[chr]; !ok || wndCnt < cnt {
				return false
			}
		}
		return true
	}

	ansL, ansR := -1, len(s)
	l := 0
	for r, chr := range s {
		wndChr2cnt[byte(chr)] += 1
		for biggerThan() {
			if r-l < ansR-ansL {
				ansL, ansR = l, r
			}
			wndChr2cnt[s[l]] -= 1
			l++
		}
	}
	if ansL < 0 {
		return ""
	}
	return s[ansL : ansR+1]
}

var MinWindow = minWindow

File: src/array/lc76.py
from typing import Dict


class Solution:
    def minWindow(self, s: str, t: str) -> str:
        tChr2cnt: Dict[str, int] = {}
        for chr in t:
            tChr2cnt[chr] = tChr2cnt.get(chr, 0) + 1
        wndChr2cnt: Dict[str, int] = {}

        def biggerThan() -> bool:
            for chr, cnt in tChr2cnt.items():
                if wndChr2cnt.get(chr, 0) < cnt:
                    return False
            return True

        ansL, ansR = -1, len(s)
        left = 0
        for right, chr in enumerate(s):
            wndChr2cnt[chr] = wndChr2cnt.get(chr, 0) + 1
            while biggerThan() and left <= right:
                if right - left < ansR - ansL:
                    ansL, ansR = left, right
                wndChr2cnt[s[left]] -= 1
                left += 1
        if ansL < 0:
            return ""
        return s[ansL : ansR + 1]
